runtime/lua/vim/iter.lua

```diff
@@ -179,6 +193,45 @@ function ListIter.filter(self, f)
   return self
 end
 
+--- Flattens an |list-iterator| pipeline, un-nesting nested values
+---
+--- Example:
+---
+--- ```lua
+--- vim.iter({1, {2}, {{3}}}):flat():totable()
+--- ```
+--- (returns {1, 2, {3}})
+---
+---@param depth number|nil Depth to which iterator pipeline should be flattened
+---                        (defaults to 1)
+---@return Iter
+function Iter.flat(self, depth) -- luacheck: no unused args
+  error('flat() requires a list-like table')
+  return self
+end
+
+---@private
+function ListIter.flat(self, depth)
+  depth = depth or 1
+  local inc = self._head < self._tail and 1 or -1
+  local target = {}
```

Comment:
I couldn't find a better way to do this than making a copy. I tried for some in-place magic, but it seemed to be going nowhere. This is basically what the ES spec does as well 🤷‍♂️